src/libraries/Microsoft.Extensions.DependencyInjection/src/ServiceLookup/StackGuard.cs

```diff
@@ -52,17 +67,15 @@ private R RunOnEmptyStackCore<R>(Func<object, R> action, object state)
                 // Using default scheduler rather than picking up the current scheduler.
                 Task<R> task = Task.Factory.StartNew(action, state, CancellationToken.None, TaskCreationOptions.DenyChildAttach, TaskScheduler.Default);
 
-                TaskAwaiter<R> awaiter = task.GetAwaiter();
-
                 // Avoid AsyncWaitHandle lazy allocation of ManualResetEvent in the rare case we finish quickly.
-                if (!awaiter.IsCompleted)
+                if (!task.IsCompleted)
                 {
                     // Task.Wait has the potential of inlining the task's execution on the current thread; avoid this.
                     ((IAsyncResult)task).AsyncWaitHandle.WaitOne();
```

Comment:
I thought we could get rid of this call all together, since Task won't run inline if there isn't sufficient space on the stack. Or was I misunderstanding?

Reply:
I _think_ we should be able to on .NET Core 3.0+, but I want to dig a little deeper and so am leaving it to be done separately.  I think it's likely less safe on .NET Framework or pre-.NET Core 3.0.  Task.Wait or Task.GetAwaiter().GetResult() here will eventually end up in TaskScheduler.TryRunInline, which will do its own depth check.  In .NET Core 3.0+, that depth check simply calls RuntimeHelpers.TryEnsureSufficientExecutionStack.  But prior to that, it was attempting to optimize away (unnecessarily) that check by first doing its own counting to see how many recursive waits were on the stack, only doing the actual depth check after ~20ish recursive waits; since the depth in this code here is likely achieved not via Task.Wait but via a deeply recursive set of service lookup calls, it'll end up assuming there's still plenty of stack left and allow the code to execute inline, defeating the purpose of this StackGuard.